### PR TITLE
Fix admin.register custom site ordering bug

### LINKED ISSUE #337: admin_site.register -> @admin.register results in "undefined name 'admin_site'"
### Python Version

3.8.13

### Django Version

4.1.7

### Package Version

1.12.0

### Description

Django-upgrade correctly transforms the `admin_site.register(StrandingSpeciesPhoto, StrandingSpeciesPhotoAdmin)` into the decorator-form`@admin.register(StrandingSpeciesPhoto, site=admin_site)`, but `admin_site` is defined after the decorator, resulting in `NameError: name 'admin_site' is not defined`.

This could be fixed by moving the definition of `admin_site` up, but I'm not sure how feasible that is. 

```diff
+ @admin.register(StrandingSpeciesPhoto, site=admin_site)
  class StrandingSpeciesPhotoAdmin(admin.ModelAdmin):
      list_display = ["__str__", "thumbnail"]
      autocomplete_fields = ["species"]
@@ -20,5 +21,4 @@ def thumbnail(self, obj):


  admin_site = StrandingAdminSite(name="stranding")
- admin_site.register(StrandingSpeciesPhoto, StrandingSpeciesPhotoAdmin)
  admin_site.register(StrandingSpecies, SpeciesAdmin)
```

### PR DESCRIPTION
**Pull request by adamchainz**
Fixes #337.

### CODE REVIEW

**Comment by adamchainz** [comment]
Currently a PoC. This approach works but is quite a lot of stuff.

Another direction we could pursue is only rewriting if the custom site is imported at the top of the module. That would be easier although it would miss some cases.

If we do adopt the approach in this PR, it would be best to make the "defined names" thing general, so we can reuse it elsewhere. Perhaps the top-level `visit()` function could track defined names, rather than `from_imports` - this would add some flexibilty to fixers.

@UnknownPlatypus do you have any input?

File: src/django_upgrade/fixers/admin_register.py [hunk 1]
```diff
@@ -36,10 +36,11 @@
 
 
 class AdminDetails:
-    __slots__ = ("parent", "model_names_per_site")
+    __slots__ = ("parent", "node", "model_names_per_site")
 
-    def __init__(self, parent: ast.AST) -> None:
+    def __init__(self, parent: ast.AST, node: ast.ClassDef) -> None:
         self.parent = parent
+        self.node = node
```
Comment (UnknownPlatypus) [inline src/django_upgrade/fixers/admin_register.py:43]: Could maybe only track `node.lineno` ? It seems like the only thing you actually use later?

**Comment by UnknownPlatypus** [comment]
> Another direction we could pursue is only rewriting if the custom site is imported at the top of the module. That would be easier although it would miss some cases.

Ye, I think tracking line number for this case might be okay, this is quite an usual case with line of codes being moved quite a lot.

> If we do adopt the approach in this PR, it would be best to make the "defined names" thing general, so we can reuse it elsewhere. Perhaps the top-level visit() function could track defined names, rather than from_imports - this would add some flexibilty to fixers.

This might add quite an overhead for a feature we rarely use (might be needed for the `MyChoices.choices` -> `MyChoices` rewrite tho). But if you see good use case / simplifications that could rely on that, it might be a good idea. Will need to handle correctly scopes and redefined names

Otherwise, maybe just a `get_FOO_defined_line` function we can sparingly use for those special cases is enough (basically what you did with `get_site_defined_line` ?

**Comment by adamchainz** [comment]
> This might add quite an overhead for a feature we rarely use (might be needed for the `MyChoices.choices` -> `MyChoices` rewrite tho). But if you see good use case / simplifications that could rely on that, it might be a good idea. Will need to handle correctly scopes and redefined names

Yeah I was thinking we'd just track module-level names, to reduce overhead of scoping and chance for bugs. Better to be a bit conservative.

**Pull request merged**